### PR TITLE
InputDialog: tap outside to close dialog if keyboard is hidden

### PR DESCRIPTION
**Pull request by hius07**
Have tried and found useful. Closes https://github.com/koreader/koreader/issues/11092.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11094)

### CODE REVIEW

File: frontend/ui/widget/inputdialog.lua [hunk 1]
```diff
@@ -478,13 +481,18 @@ function InputDialog:getAddedWidgetAvailableWidth()
     return self._input_widget.width
 end
 
--- Close the keyboard if we tap anywhere outside of the keyboard (that isn't an input field, where it would be caught via InputText:onTapTextBox)
+-- Tap outside of inputbox to hide the keyboard (inside the inputbox it is caught via InputText:onTapTextBox).
```
Comment (NiLuJe) [inline frontend/ui/widget/inputdialog.lua:484]: I don't recall what happens if we tap inside the input *dialog*, and not an input *field*. (I would assume nothing? Which is why the wording was so careful ;p).

Comment (hius07) [reply frontend/ui/widget/inputdialog.lua:484]: The keyboard becomes hidden, on the second tap - the dialog would be closed.
The reactions on tap inside the dialog (not inputfield and buttons) and outside of it are the same.

**Review by Frenzie — APPROVED** [review]
Not entirely convinced, but oh well, we'll see.

File: frontend/ui/widget/inputdialog.lua (hunk 1 above)
Comment (poire-z) [reply frontend/ui/widget/inputdialog.lua:484]: >  on the second tap - the dialog would be closed.
> The reactions on tap inside the dialog (not inputfield and buttons) and outside of it are the same.

Mhhh, this doesn't feel right to me.
If we tap outside, it is fine to close the widget.
But nowhere/never we have "tap anywhere inside a widget close the widget" (except on an obvious Close or x button).
I just tested it, and it indeed feels super odd and unexpected UX.
(OK, I don't usually tap much on a widget - but I can imagine one doing that resting his fingers, seeing the keyboard disappear, tapping again at the same place to make it reappear, and poof, lost the input text).

File: frontend/ui/widget/textviewer.lua [hunk 2]
```diff
@@ -441,6 +441,7 @@ function TextViewer:findDialog()
             {
                 {
                     text = _("Cancel"),
+                    id = "close",
```
Comment (poire-z) [inline frontend/ui/widget/textviewer.lua:444]: Remind me why the need for all these added `id="close"` ?

Comment (hius07) [reply frontend/ui/widget/textviewer.lua:444]: ButtonDialog is closed by tapping on its title ("inside").

Comment (hius07) [reply frontend/ui/widget/textviewer.lua:444]: id is used for non-touch devices and in
https://github.com/koreader/koreader/blob/082ef9b545809f865e1eeeaa8e9733a20120a1ad/frontend/ui/widget/inputdialog.lua#L635-L642

Comment (poire-z) [reply frontend/ui/widget/textviewer.lua:444]: > ButtonDialog is closed by tapping on its title ("inside"). (reply to the previous comment)

Just tested that (eg. with the dialog we get on long-press on a file in History), and it also feels odd :)

Comment (hius07) [reply frontend/ui/widget/textviewer.lua:444]: So, your proposition is
https://github.com/koreader/koreader/blob/082ef9b545809f865e1eeeaa8e9733a20120a1ad/frontend/ui/widget/confirmbox.lua#L245-L251

Comment (hius07) [reply frontend/ui/widget/textviewer.lua:444]: Or even tapping outside should not close the InputDialog? (to not lose the text)

Comment (poire-z) [reply frontend/ui/widget/textviewer.lua:444]: Yes, that's something like that I was expecting to read.

**Comment by Frenzie** [comment]
> Mhhh, this doesn't feel right to me.
If we tap outside, it is fine to close the widget.
But nowhere/never we have "tap anywhere inside a widget close the widget" (except on an obvious Close or x button).
I just tested it, and it indeed feels super odd and unexpected UX.
(OK, I don't usually tap much on a widget - but I can imagine one doing that resting his fingers, seeing the keyboard disappear, tapping again at the same place to make it reappear, and poof, lost the input text).

Agreed, I'd overlooked that.

**Comment by hius07** [comment]
Do not update ButtonDialogTitle because it's deprecated.

**Review by poire-z — APPROVED** [review]

**Comment by poire-z** [comment]
I remember I had look at that years ago, and it wasn't obvious how to solve it, or there were off side effects with some quick test implementation.
I hope it's because since then there have been many improvements/cleanup (ie. closing the keyboard, so having it not get focus) that makes this simple implementation just fine :)

File: frontend/ui/widget/inputdialog.lua (hunk 1 above)
Comment (NiLuJe) [reply frontend/ui/widget/inputdialog.lua:484]: We *should* be able to hitcheck that somewhere, though? (without checking anything: `self[1].dimen`, maybe?)

Comment (NiLuJe) [reply frontend/ui/widget/inputdialog.lua:484]: I was late to the party I see ;o).

**Comment by NiLuJe** [comment]
I had the same thoughts when approaching the PR (or even when it was mentioned in the other issue ;p), but looking at the code, I don't see any obvious traps ;).

So, yup, LGTM ;).

**Pull request merged**